react/features/base/tracks/actions.js

```diff
@@ -52,6 +54,52 @@ export function destroyLocalTracks() {
 }
 
 /**
+ * Replaces one track with another for one renegotiation instead of invoking
+ * two renegotations with a separate removeTrack and addTrack. Disposes the
+ * removed track as well.
+ *
+ * @param {JitsiLocalTrack|null} oldTrack - The track to dispose.
+ * @param {JitsiLocalTrack|null} newTrack - The track to use instead.
+ * @param {JitsiConference} [conference] - The conference from which to remove
+ * and add the tracks. If one is not provied, the conference in the redux store
+ * will be used.
+ * @returns {Function}
+ */
+export function replaceLocalTrack(oldTrack, newTrack, conference) {
+    return (dispatch, getState) => {
+        const currentConference = conference
+            || getState()['features/base/conference'].conference;
+
+        return currentConference.replaceTrack(oldTrack, newTrack)
+            .then(() => {
+                // We call dispose after doing the replace because
+                //  dispose will try and do a new o/a after the
+                //  track removes itself.  Doing it after means
+                //  the JitsiLocalTrack::conference member is already
+                //  cleared, so it won't try and do the o/a
+                const disposePromise = oldTrack
+                    ? dispatch(_disposeAndRemoveTracks([ oldTrack ]))
+                    : Promise.resolve();
+
+                return disposePromise
+                    .then(() => {
+                        if (newTrack) {
+                            const muteAction = newTrack.isVideoTrack()
+                                ? setVideoMuted : setAudioMuted;
+
+                            return dispatch(muteAction(newTrack.isMuted()));
+                        }
+                    })
+                    .then(() => {
+                        if (newTrack) {
+                            return dispatch(_addTracks([ newTrack ]));
```

Comment:
I am still thinking whether it should be the app that unmutes the track, after replace (if current track is muted, but we're adding an unmuted one). I think that the source of truth should be the app state and not state of the track being added. So if the app wants to unmute after replacing the track it should do that explicitly, by modifying app video muted state. This sync up logic adds side effect of replacing a track.